src/ui/handler/keyboard.js

```diff
@@ -125,11 +125,18 @@ class KeyboardHandler {
         return {
             cameraAnimation: (map: Map) => {
                 const zoom = map.getZoom();
+
+                // Camera is constrained to control zooming too close to terrain
+                if (map.transform._isCameraConstrained) {
+                    if (zoomDir && zoomDir > 0) zoomDir = 0;
+                    if (yDir && yDir < 0) yDir = 0;
+                    map.transform._isCameraConstrained = false;
+                }
+
```

Comment:
I'm concerned that including this logic in the keyboard handle introduces excess complexity and fragility. Could we have the same behavior on all `easeTo` and `flyTo` events (or better yet, all map movements?), allowing us to keep interactions and terrain handling separate?

Reply:
Thanks for the feedback - I'll look into `easeTo` in general (I don't think `flyTo` is used in any of these movements). There does need to be map movements that are allowed, so the user can zoom back or pan back when getting too close to terrain (otherwise they will get stuck)

Reply:
> There does need to be some map movements that are allowed, so the user can zoom back or pan back when getting too close to terrain (otherwise they will get stuck)

Would it be possible to decide if a movement is allowed or not allowed from with `transform.js`?

Do you think that we could change the behavior of `updateElevation`  `_constrainCamera()` so that they don't need any arguments and camera constraint is based on (invalid) position alone? Then instead of blocking specific interactions, invalid ones would essentially be a noop.

Or is it important for user experience that we allow zooming to a higher zoom level while constraining panning to a lower one? (The latter may well be the case, both mouse interactions seem to work well in this PR)

Besides code readability and maintainability, another reason why I think separating interaction and camera adjustment code would be preferable is that it's hard to predict how users will interact with camera position. For instance with`react-map-gl` map position is overridden at every frame (the reason is to sync position kepler.gl, but this applies to developers using the extension whether they use kepler.gl or no). So if clipping prevention depends on the specific type of interaction (or even the specific event called), we might see unpredictable behavior.